keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -292,18 +293,50 @@ is scoped to.
 or map. These will be accessible to CEL via `self`. The elements of a map or list can be validated using the CEL support for collections
 like the `all` macro, e.g. `self.all(listItem, <predicate>)` or `self.all(mapKey,
 <predicate>)`.
-  
-- For immutability use case, validator will have access to the existing version of the object. This
-  will be accessible to CEL via the `oldSelf` identifier.
-  - This will only be available on mergable collection types such as objects (unless
-    `x-kubernetes-map-type=atomic`), maps with `x-kubernetes-map-type=granular` and lists
-    with `x-kubernetes-list-type` set to `set` or `map`.  See [Merge
-    Strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) for
-    details.
-  - The use of "old" is congruent with how `AdmissionReview` identifies the existing object as
-    `oldObject`.
-  - xref [analysis of possible interactions with immutability and
-    validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable).
+
+- For certain use cases (e.g. immmutability), rules may access to the existing version of the object (see [use cases and other considerations](#access-to-existing-object-state)).
+
+  - Not all update operations are considered comparable for the purposes of this mechanism:
+
+    - Both an old and a new value must exist. It will remain possible to check if a value has been
+      added or removed by placing update rules on the parent node.
+
+    - The path to the schema node being validated must unambiguously refer to the same node -- in
+      terms of identity -- in the new object as it did in the old object. For example, array items
+      and their descendants (spec.foo[10].bar), or maps marked
+      [atomic](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy).
+
+    - Semantics from server-side apply will be enforced during CRD writes, and errors will be
+      generated if an update rule is added to a schema node that is not comparable, e.g. "<path>:
+      update rule <rule> cannot be set on schema because the schema or its parent schema is not
+      mergeable". Updates to descendants of mergeable collection types (objects not marked
+      `x-kubernetes-map-type=atomic`, maps marked `x-kubernetes-map-type=granular` and lists with
+      `x-kubernetes-list-type` set to `set` or `map`) are candidates for validation using update
+      rules.
+
+  - An optional field on rule items, "updateRule," designates a CEL expression to be evaluated
+    instead of "rule" for comparable update operations. In addition to "self", the "updateRule"
```

Comment:
> below, it says rule is still evaluated for comparable update operations

That's my mistake -- the intention was that, if provided, the "updateRule" expression alone is evaluated for comparable update operations, and in all other cases the "rule" expression alone is evaluated.

Reply:
With this intention, does @liggitt's point still stand? Would an alternative that's more explicit be viable, e.g.?

```
- rule: "..."
  operations: [CREATE]
- rule: "..."
  operations: [UPDATE]
```
(where the default is both operations)

Reply:
That's a good way to reduce ambiguity without losing support for the ratcheting use case. There's still a rough edge in that the create Rule would be evaluated on all creates, but the update Rule would not be evaluated if a given update is not comparable (e.g. field is unset in the existing object, or field belongs to an item in a list-type=map array with no item matching its key in the existing object).

What about alternative 4, which always evaluates every Rule? The kind of expression that is supported in alpha today is unchanged, but users are responsible for writing expressions that gracefully handle the absence of a comparable old object in exchange for certainty that the Rule will be evaluated.

Reply:
currently, the validation rules the user provides (both declarative and cel-based) operate only on the new object. I think the door is still open for us to auto-apply ratcheting validation by ignoring a specific bit of validation on update if the existing persisted data also fails that validation

As soon as we open the door to the user making the validation conditional, we risk losing the ability to do that.

I'm very skeptical that most users will make changes to validation rules in a way that correctly handles existing data.

Reply:
Does this case capture the problem?

`"oldSelf.x == '*' || self.x in ['A', 'B']"` (Passes if you initially set the value to '*' and then update it to 'C', but then there is no way for a ratcheting system if the 'C' should be allowed through on a subsequent update).

Looking at the use cases we've listed ("Immutability", "Prevent modification/removal once assigned" and "Append-only set") they appear semantically compatible with auto-apply ratcheting. I.e. they all test that the change to a value is safe (and not even evaluated against unchanged values). So it seems like what we're trying to make possible isn't fundamentally incompatible with auto-apply ratcheting?

Just brainstorming, but do either of these alternatives help?
- Instead of allowing for rules that are "create only" or "update only", allow for rules that only evaluate if the value is changed (and only allow these rules to have access to oldSelf)
- Don't give access to old self, but instead declare some additional functions that can be used for a subset of things we know are (a) useful and (b) safe w/r/t auto-apply ratcheting (e.g. self.isChanged())

Reply:
3. xref https://github.com/kubernetes/enhancements/pull/3160/files#r792542167

Reply:
> the user making the validation conditional

The root of the conditionality is that we can only evaluate an expression referencing existing state when it makes sense to compare it with the updated state, isn't it? If this proposal does not support comparable updates that are less strict than other writes, it's still left with some subset of validation rules that are conditionally evaluated. Is that fundamentally going to be problematic when it comes to automatic object ratcheting, or is the risk here primarily coming from having a pair of expressions: `if (comparable_update) then update_rule(old, new) else rule(new)`? @liggitt

Reply:
> As soon as we open the door to the user making the validation conditional, we risk losing the ability to do that.

if a rule attempts to use oldSelf and no oldSelf exists, I think it's a reasonable expectation to expect that rule to be skipped.

I don't see how such a choice precludes having an option on rules that do not include oldSelf to be run twice and compared for ratcheting.  A rule containing oldSelf is not a valid target for a ratcheting check.

Reply:
My long-term goal is to make CR validation smart enough to apply auto-ratcheting when existing data fails currently declared validation. This would help system controllers like namespace controller and garbage collection not get stuck.

With the current state of CRDs and CEL validation, I can imagine doing this by permitting new data to fail validation rules if existing data fails the same validation rules in the same way (with plenty of details to resolve like the level of granularity at which we do that).

As soon as we permit CEL expressions that use `oldSelf`, I'm no longer sure how to do that without breaking the expectations of the user that wrote the `oldSelf` rule. That worries me.